More matrix-free / GMG  bugs from PR 4606

@tamiko it looks like there are more bugs introduced in https://github.com/dealii/dealii/pull/4606 related to geometric multigrid and matrix-free. Can you please run the corresponding test-suite on your branch to confirm? 

The errors are convoluted but roughly read

```
lac/la_parallel_vector.h:1496:9: error: request for member ‘initialize_dof_vector’ in ‘matrix’, which is of pointer type <....>
lac/la_parallel_vector.h:1507:9: error: request for member ‘initialize_dof_vector’ in ‘matrix’, which is of pointer type <.....>
```

--- a/doc/news/changes/minor/20170714MatthiasMaier
+++ b/doc/news/changes/minor/20170714MatthiasMaier
@@ -1,0 +1,5 @@
+Improved: linear_operator gained an overload that directly takes a
+LinearOperator as exemplar object to directly copy the
+reinit_(domain|range)_vector functions from.
+<br>
+(Matthias Maier, 2017/07/14)

--- a/include/deal.II/lac/linear_operator.h
+++ b/include/deal.II/lac/linear_operator.h
@@ -1291,6 +1291,45 @@ linear_operator(const OperatorExemplar &operator_exemplar, const Matrix &matrix)
 }
 
 
+
+/**
+ * @relates LinearOperator
+ *
+ * Variant of above function that takes a LinearOperator @p
+ * operator_exemplar as an additional reference.
+ * The reinit_domain_vector and reinit_range_vector function are copied
+ * from the @p operator_exemplar object.
+ *
+ * The reference @p matrix is used to construct vmult, Tvmult, etc.
+ *
+ * This variant can, for example, be used to encapsulate preconditioners (that
+ * typically do not expose any information about the underlying matrix).
+ *
+ * @author Matthias Maier, 2017
+ *
+ * @ingroup LAOperators
+ */
+template <typename Range,
+          typename Domain,
+          typename Payload,
+          typename Matrix>
+LinearOperator<Range, Domain, Payload>
+linear_operator(const LinearOperator<Range, Domain, Payload> &operator_exemplar,
+                const Matrix &matrix)
+{
+  // Initialise the payload based on the LinearOperator exemplar
+  auto return_op = operator_exemplar;
+
+  typename std::conditional<
+  has_vmult_add_and_Tvmult_add<Range, Domain, Matrix>::type::value,
+                               MatrixInterfaceWithVmultAdd<Range, Domain, Payload>,
+                               MatrixInterfaceWithoutVmultAdd<Range, Domain, Payload>>::type().
+                               operator()(return_op, matrix);
+
+  return return_op;
+}
+
+
 //@}
 
 

--- a/include/deal.II/multigrid/mg_block_smoother.h
+++ b/include/deal.II/multigrid/mg_block_smoother.h
@@ -194,33 +194,12 @@ MGSmootherBlock<MatrixType, RelaxationType, number>::initialize (const MGMatrixT
 
   for (unsigned int i=min; i<=max; ++i)
     {
-      // Workaround: The matrix objects supplied to this class do not
-      // expose information how to reinitialize a vector. Therefore,
-      // populate vmult and Tvmult by hand...
-
-      matrices[i] = LinearOperator<BlockVector<number>>();
-      matrices[i].vmult = [&m, i](BlockVector<number> &v,
-                                  const BlockVector<number> &u)
-      {
-        m[i].vmult(v, u);
-      };
-      matrices[i].Tvmult = [&m, i](BlockVector<number> &v,
-                                   const BlockVector<number> &u)
-      {
-        m[i].Tvmult(v, u);
-      };
-
-      smoothers[i] = LinearOperator<BlockVector<number>>();
-      smoothers[i].vmult = [&s, i](BlockVector<number> &v,
-                                   const BlockVector<number> &u)
-      {
-        s[i].vmult(v, u);
-      };
-      smoothers[i].Tvmult = [&s, i](BlockVector<number> &v,
-                                    const BlockVector<number> &u)
-      {
-        s[i].Tvmult(v, u);
-      };
+      // Workaround: Unfortunately, not every "m[i]" object has a
+      // rich enough interface to populate reinit_(domain|range)_vector.
+      // Thus, apply an empty LinearOperator exemplar.
+      matrices[i] = linear_operator<BlockVector<number> >(
+                      LinearOperator<BlockVector<number> >(), m[i]);
+      smoothers[i] = linear_operator<BlockVector<number> >(matrices[i], s[i]);
     }
 }
 

--- a/include/deal.II/multigrid/mg_coarse.h
+++ b/include/deal.II/multigrid/mg_coarse.h
@@ -376,8 +376,11 @@ MGCoarseGridLACIteration<SolverType, VectorType>
   :
   solver(&s, typeid(*this).name())
 {
-  matrix = linear_operator<VectorType>(m);
-  precondition = linear_operator<VectorType>(m, p);
+  // Workaround: Unfortunately, not every "m" object has a rich enough
+  // interface to populate reinit_(domain|range)_vector. Thus, supply an
+  // empty LinearOperator exemplar.
+  matrix = linear_operator<VectorType>(LinearOperator<VectorType>(), m);
+  precondition = linear_operator<VectorType>(matrix, p);
 }
 
 
@@ -398,8 +401,11 @@ MGCoarseGridLACIteration<SolverType, VectorType>
               const PreconditionerType &p)
 {
   solver = &s;
-  matrix = linear_operator<VectorType>(m);
-  precondition = linear_operator<VectorType>(m, p);
+  // Workaround: Unfortunately, not every "m" object has a rich enough
+  // interface to populate reinit_(domain|range)_vector. Thus, supply an
+  // empty LinearOperator exemplar.
+  matrix = linear_operator<VectorType>(LinearOperator<VectorType>(), m);
+  precondition = linear_operator<VectorType>(matrix, p);
 }
 
 
@@ -433,7 +439,10 @@ void
 MGCoarseGridLACIteration<SolverType, VectorType>
 ::set_matrix(const MatrixType &m)
 {
-  matrix = linear_operator<VectorType>(m);
+  // Workaround: Unfortunately, not every "m" object has a rich enough
+  // interface to populate reinit_(domain|range)_vector. Thus, supply an
+  // empty LinearOperator exemplar.
+  matrix = linear_operator<VectorType>(LinearOperator<VectorType>(), m);
 }
 
 DEAL_II_ENABLE_EXTRA_DIAGNOSTICS

--- a/include/deal.II/multigrid/mg_matrix.h
+++ b/include/deal.II/multigrid/mg_matrix.h
@@ -185,20 +185,11 @@ namespace mg
     matrices.resize(p.min_level(), p.max_level());
     for (unsigned int level = p.min_level(); level <= p.max_level(); ++level)
       {
-        // Workaround: The matrix objects supplied to this class do not
-        // expose information how to reinitialize a vector. Therefore,
-        // populate vmult and Tvmult by hand...
-        matrices[level] = LinearOperator<VectorType>();
-        matrices[level].vmult = [&p, level](
-                                  VectorType &v, const VectorType &u)
-        {
-          p[level].vmult(v, u);
-        };
-        matrices[level].Tvmult = [&p, level](
-                                   VectorType &v, const VectorType &u)
-        {
-          p[level].Tvmult(v, u);
-        };
+        // Workaround: Unfortunately, not every "p[level]" object has a
+        // rich enough interface to populate reinit_(domain|range)_vector.
+        // Thus, apply an empty LinearOperator exemplar.
+        matrices[level] =
+          linear_operator<VectorType>(LinearOperator<VectorType>(), p[level]);
       }
   }
 

--- a/include/deal.II/multigrid/mg_smoother.h
+++ b/include/deal.II/multigrid/mg_smoother.h
@@ -1036,7 +1036,7 @@ MGSmootherPrecondition<MatrixType, PreconditionerType, VectorType>::initialize
 
   for (unsigned int i=min; i<=max; ++i)
     {
-      matrices[i] = &m[i];
+      matrices[i] = linear_operator<VectorType>(m[i]);
       smoothers[i].initialize(m[i], data[i]);
     }
 }

--- a/include/deal.II/multigrid/mg_smoother.h
+++ b/include/deal.II/multigrid/mg_smoother.h
@@ -805,7 +805,11 @@ MGSmootherRelaxation<MatrixType, RelaxationType, VectorType>::initialize
 
   for (unsigned int i=min; i<=max; ++i)
     {
-      matrices[i] = linear_operator<VectorType>(m[i]);
+      // Workaround: Unfortunately, not every "m[i]" object has a rich
+      // enough interface to populate reinit_(domain|range)_vector. Thus,
+      // apply an empty LinearOperator exemplar.
+      matrices[i] =
+        linear_operator<VectorType>(LinearOperator<VectorType>(), m[i]);
       smoothers[i].initialize(m[i], data);
     }
 }
@@ -830,7 +834,11 @@ MGSmootherRelaxation<MatrixType, RelaxationType, VectorType>::initialize
 
   for (unsigned int i=min; i<=max; ++i)
     {
-      matrices[i] = linear_operator<VectorType>(m[i]);
+      // Workaround: Unfortunately, not every "m[i]" object has a rich
+      // enough interface to populate reinit_(domain|range)_vector. Thus,
+      // apply an empty LinearOperator exemplar.
+      matrices[i] =
+        linear_operator<VectorType>(LinearOperator<VectorType>(), m[i]);
       smoothers[i].initialize(m[i], data[i]);
     }
 }
@@ -852,7 +860,11 @@ MGSmootherRelaxation<MatrixType, RelaxationType, VectorType>::initialize
 
   for (unsigned int i=min; i<=max; ++i)
     {
-      matrices[i] = &(m[i].block(row, col));
+      // Workaround: Unfortunately, not every "m[i]" object has a rich
+      // enough interface to populate reinit_(domain|range)_vector. Thus,
+      // apply an empty LinearOperator exemplar.
+      matrices[i] = linear_operator<VectorType>(LinearOperator<VectorType>(),
+                                                m[i].block(row, col));
       smoothers[i].initialize(m[i].block(row, col), data);
     }
 }
@@ -879,7 +891,11 @@ MGSmootherRelaxation<MatrixType, RelaxationType, VectorType>::initialize
 
   for (unsigned int i=min; i<=max; ++i)
     {
-      matrices[i] = &(m[i].block(row, col));
+      // Workaround: Unfortunately, not every "m[i]" object has a rich
+      // enough interface to populate reinit_(domain|range)_vector. Thus,
+      // apply an empty LinearOperator exemplar.
+      matrices[i] = linear_operator<VectorType>(LinearOperator<VectorType>(),
+                                                m[i].block(row, col));
       smoothers[i].initialize(m[i].block(row, col), data[i]);
     }
 }
@@ -1009,7 +1025,11 @@ MGSmootherPrecondition<MatrixType, PreconditionerType, VectorType>::initialize
 
   for (unsigned int i=min; i<=max; ++i)
     {
-      matrices[i] = linear_operator<VectorType>(m[i]);
+      // Workaround: Unfortunately, not every "m[i]" object has a rich
+      // enough interface to populate reinit_(domain|range)_vector. Thus,
+      // apply an empty LinearOperator exemplar.
+      matrices[i] =
+        linear_operator<VectorType>(LinearOperator<VectorType>(), m[i]);
       smoothers[i].initialize(m[i], data);
     }
 }
@@ -1036,7 +1056,11 @@ MGSmootherPrecondition<MatrixType, PreconditionerType, VectorType>::initialize
 
   for (unsigned int i=min; i<=max; ++i)
     {
-      matrices[i] = linear_operator<VectorType>(m[i]);
+      // Workaround: Unfortunately, not every "m[i]" object has a rich
+      // enough interface to populate reinit_(domain|range)_vector. Thus,
+      // apply an empty LinearOperator exemplar.
+      matrices[i] =
+        linear_operator<VectorType>(LinearOperator<VectorType>(), m[i]);
       smoothers[i].initialize(m[i], data[i]);
     }
 }

